packages/react/src/__tests__/state.test.tsx

```diff
@@ -0,0 +1,273 @@
+import { act, fireEvent, render } from '@testing-library/react'
+import React from 'react'
+
+import { event } from '../events'
+import { DataSource, Template } from '../refs'
+import { response } from '../response'
+import { increment, setStateAction, state, toggle } from '../state'
+import { BUILT_IN_COMPONENTS } from './template.test'
+
+it('should render initial state', () => {
+  const value = state('foo')
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: () => jest.fn(),
+      get() {
+        return {
+          $: 'component',
+          component: 'View',
+          children: value,
+        }
+      },
+    }),
+    sendEvent: jest.fn(),
+  }
+  const component = render(<Template components={BUILT_IN_COMPONENTS} dataSource={dataSource} />)
+  expect(component.asFragment()).toMatchInlineSnapshot(`
+    <DocumentFragment>
+      <div>
+        foo
+      </div>
+    </DocumentFragment>
+  `)
+})
+
+it('should set state with default payload', async () => {
+  const value = state('foo')
+  const dataSource: DataSource = {
+    get: () => ({
+      subscribe: () => jest.fn(),
+      get() {
+        return [
+          {
+            $: 'component',
+            key: 'header',
+            component: 'View',
+            children: value,
+          },
+          {
+            $: 'component',
+            key: 'button',
+            component: 'View',
+            props: {
+              ['data-testid']: 'button',
+              onClick: event(setStateAction(value)),
```

Comment:
I'm still not convinced the user should need to wrap with `event()` here

Reply:
When you use JSX syntax, you don't have to. Haven't configured this in the unit test, so I am doing it explicitly.

Reply:
hmm. but I wonder about the actual JSON definition, it should be fine to provide an action directly as a handler, right?

Reply:
So actual JSON definition allows for two things:
```
{$: 'event', handler: () => {}, actions: []}
```
^ an event that has function to execute on the server
and
```
{$: 'actions', actions: []}
```
like event, but is local only.

When you call `event(something)`, that helper chooses which thing to return based on arguments. If you do `event(action)` (no handler), it returns `{$: 'actions', actions: []}`. If you do `event(fn)`, it does the former. 

We chose not to support passing a single action or an array of actions, to avoid looping over an entire array to make sure all elements are of the same type. Instead, users have to wrap them with `{$: 'actions', actions: []}`.

Reply:
We could allow passing a single action, or an array of actions (requires checking each element of an array is of the same type), now that I think about it.

We would then, on the client, essentially wrap them with `{$: 'actions', actions: []}`, since that's already what our code expects.

Reply:
hmm, well a single action doesn't need the array or `$: 'actions'`

Reply:
Yeah we can always add support for passing single action. But it's only going to be useful for those not using JSX and using raw JSON.